### ISSUE TITLE
[Question] debug mxnet op with gdb

### ISSUE THREAD

**Issue by taras-sereda** [dmlc/mxnet#4881]
Question: How one could debug mxnet operators using python bindings? How do you test new operators while developing them? Would be happy to hear some advice and best practices. 
I have simple one op code, with a single convolution. I would like to step in the code of Forward pass. What should be done to do that?

I've tried the method described bellow, so far it brings no fruits.  If there are some mistakes or misunderstanding would be happy if you can point on them. 

## Environment info
Operating System: MacOS Siera
Compiler: gcc
Package used : Python
MXNet version: 0.9.3
Python version and distribution: python2.7
## Minimum reproducible example
```
import mxnet as mx

data_shape = (1,3,5,5)
class SimpleData(object):

    def __init__(self, data):
        self.data = data

data = mx.sym.Variable('data')
conv = mx.sym.Convolution(data=data, kernel=(3,3), pad=(1,1), stride=(1,1), num_filter=1)

mod = mx.mod.Module(conv)
mod.bind(data_shapes=[('data', data_shape)])
mod.init_params()

input_data = mx.nd.ones(data_shape)
mod.forward(data_batch=SimpleData([input_data]))
res = mod.get_outputs()[0].asnumpy()
print(res)
```
## Steps to reproduce
1. mxnet built with DEBUG=1
2. expected behavior: execution should stop on specified break point, however it continues to run. 
```
taras@~/PycharmProjects/learn_MXNet$>gdb --args python  debug_conv.py
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin16.3.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...(no debugging symbols found)...done.
(gdb) b ConvolutionOp::Forward
Function "ConvolutionOp::Forward" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (ConvolutionOp::Forward) pending.
(gdb) run
Starting program: /usr/local/bin/python debug_conv.py
[New Thread 0x1403 of process 45783]
warning: unhandled dyld version (15)
[New Thread 0x1503 of process 45783]

Thread 3 received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 0x1503 of process 45783]
0x0000000100005000 in ?? ()
(gdb) c
Continuing.
[[[[ 0.02968991  0.03496315  0.03496315  0.03496315  0.01793109]
   [ 0.04184391  0.04823352  0.04823352  0.04823352  0.02337554]
   [ 0.04184391  0.04823352  0.04823352  0.04823352  0.02337554]
   [ 0.04184391  0.04823352  0.04823352  0.04823352  0.02337554]
   [ 0.02533177  0.03832858  0.03832858  0.03832858  0.01590874]]]]
[Inferior 1 (process 45783) exited normally]
(gdb)
```

**Comment by piiswrong** [comment]
you can use monitor (example/python-howto/monitor_weight) and put pdb breakpoint in stat function

**Comment by feiyulv** [comment]
It seems the breakpoint  you set doesn't work. Usually I set break point to file:line, that will work, like this.
```
b src/operator/convolution-inl.h:100
```

**Comment by taras-sereda** [comment]
thank you @feiyulv , I've tried the way you propose. 
both using absolute and relative path to the source file.
```
(gdb) b src/operator/convolution-inl.h:109
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (src/operator/convolution-inl.h:109) pending.
(gdb) b /Users/taras/machine_learning/mxnet/src/operator/convolution-inl.h:109
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (/Users/taras/machine_learning/mxnet/src/operator/convolution-inl.h:109) pending.
(gdb) i b
Num     Type           Disp Enb Address    What
1       breakpoint     keep y   <PENDING>  src/operator/convolution-inl.h:109
2       breakpoint     keep y   <PENDING>  /Users/taras/machine_learning/mxnet/src/operator/convolution-inl.h:109

```
still the execution doesn't stop at the break points specified. I think the problem could be that symbol tables are not loaded for the functions I would like to debug. Why is it so, I don't understand. 

Were you successful debugging this particular example? I've tried to debug with gdb other libraries, setting up break points both per method level or row level - it worked fine. Would be happy to debug this way mxnet as well.

**Comment by feiyulv** [comment]
@taras-sereda Do you use cudnn? If  cudnn flag is true, the breakpoint you set will not be run, since it will use cudnn to do convolution. You can print a info after this line to confirm it.

**Comment by taras-sereda** [comment]
@feiyulv I'm building the library without CUDA and CUDNN. I've added print line to 
```src/operator/convolution-inl.h``` and can confirm that I use indeed this source. May be it's necessary to switch off multi threaded execution to have more control over execution process. Does anybody know how to run mxnet in single thread mode?
```
Starting program: /usr/local/bin/python debug_conv.py
[New Thread 0x1403 of process 57061]
warning: unhandled dyld version (15)
[New Thread 0x1503 of process 57061]

Thread 3 received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 0x1503 of process 57061]
```

**Comment by taras-sereda** [comment]
Finally I was able to hit the specified break point. Before that I was trying to debug on MacOS, now I did the same on Ubuntu. I still don't know why debugging under MacOS brought no results, at least Ubuntu didn't let me down. 
@feiyulv thank you for hints and discussion.

**Comment by bigdata2** [comment]
@taras-sereda did you do anything different to make it hit breakpoint on Ubuntu?

 I am unable to hit breakpoint using your example above either from OS X (Sierra) or Ubuntu 14.04. In addition, I do not see test std::cout statements that I put in Backward, Forward functions as well as in ConvolutionOp constructor. I compiled the code using DEBUG = 1 and can see -g option passed during compilation. I also copied built libmxnet.so into /usr/local/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet before invoking python with gdb.
